### PR TITLE
feat (#major); goldfinch; add goldfinch subgraph

### PR DESCRIPTION
**Pull request by tnkrxyz**
## TODO
- [x] README & Methodology
- [x] testing deployment
- [x] QA 
   - test deployment: https://api.thegraph.com/subgraphs/name/tnkrxyz/benqi (the url is correct & sorry have to re-use the handle)
   - validation dashboard: https://subgraphs.messari.io/subgraph?endpoint=https://api.thegraph.com/subgraphs/name/tnkrxyz/benqi&tab=protocol

### CODE REVIEW

**Comment by dmelotik** [comment]
Make sure to update the file names to match the same scheme as other .ts files ✅

**Comment by tnkrxyz** [comment]
> Make sure to update the file names to match the same scheme as other .ts files ✅

This is something we have to decide sometime down the road, because almost all file names are from goldfinch's official subgraph. We made the decision to merge our schema with theirs for this iteration and we will decide later whether it makes sense to continue this way.

File: subgraphs/goldfinch/src/mappings/tranched_pool.ts
Comment (dmelotik) [inline subgraphs/goldfinch/src/mappings/tranched_pool.ts:136]: shorten up this line by replacing `protocol._marketIDs!.length` w/ `marketIDs.length`

File: subgraphs/goldfinch/src/mappings/tranched_pool.ts
Comment (dmelotik) [inline subgraphs/goldfinch/src/mappings/tranched_pool.ts:149]: seems like a development log we can remove

File: subgraphs/goldfinch/src/mappings/tranched_pool.ts
Comment (dmelotik) [inline subgraphs/goldfinch/src/mappings/tranched_pool.ts:191]: remove if not being used

File: subgraphs/goldfinch/src/mappings/tranched_pool.ts
Comment (dmelotik) [inline subgraphs/goldfinch/src/mappings/tranched_pool.ts:243]: is goldfinch guarunteed to only be USDC pools?

File: subgraphs/goldfinch/src/mappings/tranched_pool.ts
Comment (dmelotik) [inline subgraphs/goldfinch/src/mappings/tranched_pool.ts:337]: remove if unused

File: subgraphs/goldfinch/src/mappings/tranched_pool.ts
Comment (dmelotik) [inline subgraphs/goldfinch/src/mappings/tranched_pool.ts:384]: same

File: subgraphs/goldfinch/src/mappings/tranched_pool.ts
Comment (dmelotik) [inline subgraphs/goldfinch/src/mappings/tranched_pool.ts:465]: same

File: subgraphs/goldfinch/src/mappings/tranched_pool.ts
Comment (dmelotik) [inline subgraphs/goldfinch/src/mappings/tranched_pool.ts:472]: same

File: subgraphs/goldfinch/src/entities/tranched_pool.ts
Comment (dmelotik) [inline subgraphs/goldfinch/src/entities/tranched_pool.ts:256]: BIGINT_ZERO

File: subgraphs/goldfinch/src/entities/tranched_pool.ts
Comment (dmelotik) [inline subgraphs/goldfinch/src/entities/tranched_pool.ts:334]: remove if unnecessary for our case

File: subgraphs/goldfinch/src/mappings/senior_pool.ts
Comment (dmelotik) [inline subgraphs/goldfinch/src/mappings/senior_pool.ts:74]: remove this

File: subgraphs/goldfinch/src/mappings/senior_pool.ts
Comment (dmelotik) [inline subgraphs/goldfinch/src/mappings/senior_pool.ts:1]: A lot of the things to update in this contract are the same things I have said in others

**Review by dmelotik — COMMENTED** [review]
My argument for keeping messari subgraph standards is that since the subgraph is now in our repo, it follows our standard schema, it should also follow our development practices. Will make it easier to maintain in the long term.

That would mean changing the file names and added `_` to any additional entities added to our schema.

Some of the pools are negative?

Since we want lending rewards to be good now, it looks like we will need to update the mappings to get it working. Looks like we have an amount but no price. You can derive price from a uniswap LP

**Comment by jaimehgb** [comment]
> My argument for keeping messari subgraph standards is that since the subgraph is now in our repo, it follows our standard schema, it should also follow our development practices. Will make it easier to maintain in the long term.
> 
> That would mean changing the file names and added _ to any additional entities added to our schema.

In the specific case of goldfinch, we decided to have both schemas merged (ours and theirs). So I think we can make the exception. We use the `_` to indicate a field or entity is auxiliary, or not part of the schema, but we could say we are using some extended version for goldfinch.

File: subgraphs/goldfinch/src/mappings/tranched_pool.ts
Comment (dmelotik) [reply subgraphs/goldfinch/src/mappings/tranched_pool.ts:118]: @tnkrxyz even though this was part of their subgraph I think we should fix this since it is causing incorrect numbers. thoughts?

**Comment by tnkrxyz** [comment]
> 



> Apologies for the spam here, I was confused why we were doing it like this, but now I am enlightened. I resolved most of my comments, but there are still a couple things I wanted to point out.

np!
 
> Also, since we want lending rewards to be good now, it looks like we will need to update the mappings to get it working. Looks like we have an amount but no price. You can derive price from a uniswap LP

Sure, I will add GFI price & rewardTokenEmissionUSD

File: subgraphs/goldfinch/README.md
Comment (tnkrxyz) [reply subgraphs/goldfinch/README.md:74]: Makes sense. I will remove the Build section.

File: subgraphs/goldfinch/README.md
Comment (tnkrxyz) [reply subgraphs/goldfinch/README.md:94]: Will do.

File: subgraphs/goldfinch/src/common/getters.ts
Comment (tnkrxyz) [reply subgraphs/goldfinch/src/common/getters.ts:166]: I reset isActive and canBorrowFrom when a market is created & after checking their borrowing limit in senior_pool.ts and tranched_pool.ts

File: subgraphs/goldfinch/src/common/helpers.ts
Comment (tnkrxyz) [reply subgraphs/goldfinch/src/common/helpers.ts:175]: The switch is here for senior_pool to avoid double counting revenues collected from tranched pools: https://github.com/tnkrxyz/subgraphs/blob/a8198bd8ac4c278e913073b224d813651e5afbd6/subgraphs/goldfinch/src/mappings/senior_pool.ts#L387-L399

File: subgraphs/goldfinch/src/mappings/tranched_pool.ts
Comment (tnkrxyz) [reply subgraphs/goldfinch/src/mappings/tranched_pool.ts:119]: I believe the negative TVL is due to incorrect handle of migrated pools (deposits were not recorded) and is fixed with https://github.com/messari/subgraphs/pull/1298/commits/a8198bd8ac4c278e913073b224d813651e5afbd6 . The official subgraph has a way to get principle deposited to tranched pools, but seems too complicated: https://github.com/tnkrxyz/subgraphs/blob/a8198bd8ac4c278e913073b224d813651e5afbd6/subgraphs/goldfinch/src/entities/tranched_pool.ts#L145-L212 . From what I see summing deposit amounts give reasonable results.

File: subgraphs/goldfinch/src/common/helpers.ts
Comment (tnkrxyz) [reply subgraphs/goldfinch/src/common/helpers.ts:491]: Good catch/good question. I need to think about this more. For goldfinch, a depositor can add/reduce/close an open position, but for tranched pools, a depositor can not add to an open position, they can only reduce or close it. Borrowers can add, reduce, or close their position.

**Comment by dmelotik** [comment]
@tnkrxyz should we fix the lint issues or do we want it to remain the same as Goldfinch's subgraph and override

**Comment by tnkrxyz** [comment]
> @tnkrxyz should we fix the lint issues or do we want it to remain the same as Goldfinch's subgraph and override

I will fix the lint issues. 

I am working on addressing the QA issues in https://github.com/messari/subgraphs/issues/1409 and plan to address them before this PR is merged. Does it make sense? Or should we merge this one first & start another PR for QA?

**Comment by dmelotik** [comment]
> I am working on addressing the QA issues in https://github.com/messari/subgraphs/issues/1409 and plan to address them before this PR is merged. Does it make sense? Or should we merge this one first & start another PR for QA?

If it is a lot of changes and this PR is pretty much good to go then let's merge, but if it is small let's add it in. Let me know what you want to do, and I will review accordingly.

**Comment by tnkrxyz** [comment]
> 



> > I am working on addressing the QA issues in #1409 and plan to address them before this PR is merged. Does it make sense? Or should we merge this one first & start another PR for QA?
> 
> If it is a lot of changes and this PR is pretty much good to go then let's merge, but if it is small let's add it in. Let me know what you want to do, and I will review accordingly. @tnkrxyz

Thanks for the input! I am almost done with the QA fixes, so I'll add them to this PR.

**Comment by ishraq8** [comment]
Addition to the docs and QA look good to go.

@dmelotik said it was good to go as well, but i'll let him have the last word before merge.

**Review by dmelotik — APPROVED** [review]
Nice work!

I am approving, but 2 things to fix:
- Change the status to `prod`
- And there is a missing check for reverted function call that is failing to index: 
Screen Shot 2022-12-07 at 1 08 48 PM

File: subgraphs/goldfinch/src/common/helpers.ts
Comment (tnkrxyz) [reply subgraphs/goldfinch/src/common/helpers.ts:503]: I am closing this and make a note in #1499 to look into this.

**Comment by tnkrxyz** [comment]
@dmelotik , where did you see the error? This is caused by goldfinch deployed a new senior pool contract deprecated compoundBalance function. Since this is in their code & I hesitate to make substantial changes, so I added a bypass past the new senior pool deployment block: https://github.com/tnkrxyz/subgraphs/blob/847f33e3f09faadedaa797734612007f414bcfb3/subgraphs/goldfinch/src/entities/senior_pool.ts#L92 . The  deployment I make is not having this error: https://thegraph.com/hosted-service/subgraph/tnkrxyz/goldfinch

**Comment by dmelotik** [comment]
Hey @tnkrxyz, looking at that again, to avoid this in case they ever deploy more pools like this, you should use the (`try_call()` if reverted == 0) logic.

In fact it should probably be used in that whole file (I missed it). The `getOrElse()` function probably fits best there.

**Comment by tnkrxyz** [comment]
> 



> Hey @tnkrxyz, looking at that again, to avoid this in case they ever deploy more pools like this, you should use the (`try_call()` if reverted == 0) logic.
> 
> In fact it should probably be used in that whole file (I missed it). The `getOrElse()` function probably fits best there.

Again, I am reluctant to change code from the Goldfinch official subgraph, as they may make change on their repository and we may have to merge their changes into ours at some point or vice versa.

**Pull request merged**
